s2a/src/main/java/io/grpc/s2a/handshaker/S2AStub.java

```diff
@@ -0,0 +1,221 @@
+import java.util.concurrent.BlockingQueue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/** Reads and writes messages to and from the S2A. */
+@NotThreadSafe
+class S2AStub implements AutoCloseable {
+  private static final Logger logger = Logger.getLogger(S2AStub.class.getName());
+  private static final long HANDSHAKE_RPC_DEADLINE_SECS = 20;
+  private final StreamObserver<SessionResp> reader = new Reader();
+  private final BlockingQueue<Result> responses = new ArrayBlockingQueue<>(10);
+  private S2AServiceGrpc.S2AServiceStub serviceStub;
+  private StreamObserver<SessionReq> writer;
+  private boolean doneReading = false;
+  private boolean doneWriting = false;
+
+  static S2AStub newInstance(S2AServiceGrpc.S2AServiceStub serviceStub) {
+    checkNotNull(serviceStub);
+    return new S2AStub(serviceStub);
+  }
+
+  @VisibleForTesting
+  static S2AStub newInstanceForTesting(StreamObserver<SessionReq> writer) {
+    checkNotNull(writer);
+    return new S2AStub(writer);
+  }
+
+  private S2AStub(S2AServiceGrpc.S2AServiceStub serviceStub) {
+    this.serviceStub = serviceStub;
+  }
+
+  private S2AStub(StreamObserver<SessionReq> writer) {
+    this.writer = writer;
+  }
+
+  @VisibleForTesting
+  StreamObserver<SessionResp> getReader() {
+    return reader;
+  }
+
+  @VisibleForTesting
+  BlockingQueue<Result> getResponses() {
+    return responses;
+  }
+
+  /**
+   * Sends a request and returns the response. Caller must wait until this method executes prior to
+   * calling it again. If this method throws {@code ConnectionClosedException}, then it should not
+   * be called again, and both {@code reader} and {@code writer} are closed.
+   *
+   * @param req the {@code SessionReq} message to be sent to the S2A server.
+   * @return the {@code SessionResp} message received from the S2A server.
+   * @throws ConnectionClosedException if {@code reader} or {@code writer} calls their {@code
+   *     onCompleted} method.
+   * @throws IOException if an unexpected response is received, or if the {@code reader} or {@code
+   *     writer} calls their {@code onError} method.
+   */
+  public SessionResp send(SessionReq req) throws IOException, InterruptedException {
+    if (doneWriting && doneReading) {
+      logger.log(Level.INFO, "Stream to the S2A is closed.");
+      throw new ConnectionClosedException("Stream to the S2A is closed.");
+    }
+    createWriterIfNull();
+    if (!responses.isEmpty()) {
+      IOException exception = null;
+      SessionResp resp = null;
+      try {
+        resp = responses.take().getResultOrThrow();
+      } catch (IOException e) {
+        exception = e;
+      }
+      responses.clear();
+      if (exception != null) {
+        throw new IOException(
+            "Received an unexpected response from a host at the S2A's address. The S2A might be"
+                + " unavailable."
+                + exception.getMessage());
```

Comment:
This should be the cause of the IOException, not just add the message. Adding the message loses the exception type, the stack trace, and any causes that `exception` has. `new IOException("Received an ... unavailable", exception)`.

Reply:
Thanks for pointing this out, done in https://github.com/grpc/grpc-java/pull/11570